### PR TITLE
MPI-4: Add support for MPI Sessions

### PR DESCRIPTION
**Pull request by hppritcha**
The PR contains changes to Open MPI to support the MPI 4 Sessions API (see chapter 11 sections 3 and 4, and chapter 7 sections 3.2 and 4.2, and parts of chapter 9 concerning errorhandlers).  The MPI 4 standard is available at https://www.mpi-forum.org/docs/mpi-4.0/mpi40-report.pdf.

A set of sessions tests has been added to the OMPI public tests repo:https://github.com/open-mpi/ompi-tests-public.
A set of issues concerning this PR are at https://github.com/hpc/ompi/issues

Major changes associated with this PR are:

- replacement of a single ompi/opal core initialize/finalize with a notion of instances and use of destructor callbacks to clean up rather than relying on an instance finalize procedure tightly coupled with the instance initialization procedure
-  use of an extended CID (exCID) concept for handling Communicator IDs (this is conditional depending on whether a PML supports this).  The OB1 and CM/OFI PML's have been modified as part of this PR to support exCIDs.
- support for calling functions included in Table 11.1 of the MPI 4 standard before calling ```MPI_Init``` and friends and after calling ```MPI_Finalize``` and friends.

A paper describing the prototype (at least at an early stage) is available at https://ieeexplore.ieee.org/abstract/document/8891002

There will be extensive squashing of commits prior to merging so it would probably be best to focus on the changes introduced by this PR as a whole rather than reviewing individual commits.

### CODE REVIEW

File: examples/ring_c.c
Comment (bosilca) [inline examples/ring_c.c:10]: Why is this beautification change part of the sessions PR ?

File: ompi/attribute/attribute.c
Comment (bosilca) [inline ompi/attribute/attribute.c:414]: Why is this necessary ? As far as I see there is still a single global object to handle the attributes ?

File: ompi/attribute/attribute.c
Comment (bosilca) [inline ompi/attribute/attribute.c:543]: Indentations are inconsistent.

File: ompi/communicator/comm.c
Comment (bosilca) [inline ompi/communicator/comm.c:210]: This is one sneaky change, with drastic implications in terms of latency, scalability and memory usage. It need to be carefully assessed and discusses before being merged.

File: ompi/communicator/comm.c
Comment (bosilca) [inline ompi/communicator/comm.c:2408]: This thread-local 2 correct values function begs for documentation !

File: ompi/communicator/comm_cid.c
Comment (bosilca) [inline ompi/communicator/comm_cid.c:337]: How scalable is the construction of the PMIx group ? What process is calling this function ?

File: ompi/communicator/communicator.h
Comment (bosilca) [inline ompi/communicator/communicator.h:142]: Why the name change ? The old name was clear enough.

File: ompi/communicator/comm_cid.c
Comment (rhc54) [reply ompi/communicator/comm_cid.c:337]: It uses the same PRRTE communication collective as the "modex" operation. All processes that are to be members of the group must call it. It basically is a data-less "modex" (aka barrier) that returns a unique context ID.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

**Comment by hppritcha** [comment]
bot:ibm:retest

File: ompi/communicator/communicator.h
Comment (hppritcha) [reply ompi/communicator/communicator.h:142]: @hjelmn any reason not to go back to ompi_mpi_communicators?

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

File: ompi/communicator/comm.c
Comment (hppritcha) [reply ompi/communicator/comm.c:210]: yes agreed.   There's an issue to address this in a different way - https://github.com/hpc/ompi/issues/65

File: examples/ring_c.c
Comment (hppritcha) [reply examples/ring_c.c:10]: I'll put back the ring_c.c from master.  I think @hjelmn was doing some clang formatting here.

**Comment by hppritcha** [comment]
bot:ibm:retest

File: ompi/communicator/communicator.h
Comment (hjelmn) [reply ompi/communicator/communicator.h:142]: I can't remember why I changed the name. There probably was a reason at one point during the evolution of this branch. Most likely I was making the name consistent with ```ompi_comm_hash``` to make it clear that one is an array of communicators where the lookup is by index where the other one is a hash where the lookup is by exCID.

File: ompi/attribute/attribute.c
Comment (hjelmn) [reply ompi/attribute/attribute.c:414]: Looks like it was done to use opal-object reference counting. The last one to tear down attributes calls the destructor. At some point Open MPI should really stop using globals and and this change is a step in the right direction.

File: ompi/attribute/attribute.c
Comment (hjelmn) [reply ompi/attribute/attribute.c:543]: :+1: maybe tabs vs spaces? Also, maybe avoid using goto here either by putting the contents between the lock/unlock into a seperate function or use something like ```do { } while(false);```

File: ompi/communicator/comm.c
Comment (hjelmn) [reply ompi/communicator/comm.c:2408]: Indeed. I thought I documented this somewhere.... Guess not. Should add documentation to the header declaration.

File: ompi/attribute/attribute.c
Comment (bosilca) [reply ompi/attribute/attribute.c:414]: Im not arguing against avoiding globals, but  such a change deserve its own PR instead of being part of a very large one.

File: ompi/communicator/comm_cid.c
Comment (bosilca) [reply ompi/communicator/comm_cid.c:337]: In other words, slightly less scalable than MPI collectives on the local costs (because of the need to construct and destruct the group), then delegated to a different software that uses some communication capabilities, to finally be delegated to a RM with or without efficient support for collectives. Hopefully this CID selection is only rarely used and most calls are outside the critical path. Right ?

File: ompi/communicator/communicator.h
Comment (bosilca) [reply ompi/communicator/communicator.h:142]: If we remove this name change, the PR is getting more self-contained (aka. smaller and easier to understand).

Comment (hjelmn) [reply ompi/communicator/communicator.h:142]: SGTM

File: examples/ring_c.c
Comment (hjelmn) [reply examples/ring_c.c]: No. My guess is that this happened when merging with master. I think the ordering on master changed when the examples were formatted. Anyway, looked like it is fixed now so hopefully there are not any more changes like that. Will spot check and see.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/2175efaa13b971835186c294806b0001

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/61a0a9029bfafb0c95f43f60957a65d5

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/52e26384a669f512dad96ba55ceaa81f

**Comment by hppritcha** [comment]
hmm aws seems to have gone bonkers rerun that first
bot:aws:retest

**Comment by hppritcha** [comment]
I suspect the git commit checker is experiencing some kind of scalability melt down.  will ignore going forward with this PR. @jsquyres

**Comment by jsquyres** [comment]
> I suspect the git commit checker is experiencing some kind of scalability melt down. will ignore going forward with this PR. @jsquyres

It looks like the git commit checker is not running at all.  I've seen this before -- rarely, but I have seen it.  I haven't been able to figure out the root cause.  Perhaps it doesn't run if there's a conflict on the PR...?

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/c6fe74035df169e07a5eb531a01cf04d

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/77e50f876685171598bd552c135adaf2

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/358a6ff4065920a662e408d5032d02e7

File: ompi/communicator/communicator.h
Comment (hppritcha) [reply ompi/communicator/communicator.h]: switched back to the old variable name

**Comment by hppritcha** [comment]
Leaving all the commits as is for now.  Will squash down to a much smaller set of commits once this PR is reviewd.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/5993530c5eb1eb4702834c6595aef513

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/1f5d55073158a8481863651d7350898a

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/6e13aca998586ea4dd119e1c8a7a6981

File: ompi/attribute/attribute.c
Comment (hjelmn) [reply ompi/attribute/attribute.c:414]: No argument there. Up to @hppritcha to decide whether to move it to another PR.

File: ompi/attribute/attribute.c
Comment (hjelmn) [inline ompi/attribute/attribute.c:658]: For consistency this should probably be OMPI_SUCCESS.

File: ompi/attribute/attribute.c
Comment (hjelmn) [inline ompi/attribute/attribute.c:800]: Same here. OMPI_SUCCESS for consistency. These functions should probably all work with internal error codes that are then translated to the MPI error space in the bindings.

Maybe we should split out all the changes to these files into a PR now. Would be easier to review. I can do that if you like. I have some time today to work on OMPI stuff.

Comment (hppritcha) [reply ompi/attribute/attribute.c:800]: that would be great Nathan.  thanks!

**Comment by hppritcha** [comment]
okay let's not do any more of this if we can.  this is a mess to rebase

File: opal/runtime/opal_finalize.c
Comment (hjelmn) [inline opal/runtime/opal_finalize.c:117]: we can drop this change.

File: opal/mca/btl/sm/btl_sm_fifo.h
Comment (hjelmn) [inline opal/mca/btl/sm/btl_sm_fifo.h:161]: We can drop this change.

File: ompi/mpi/c/profile/Makefile.am
Comment (hjelmn) [inline ompi/mpi/c/profile/Makefile.am:1]: Do you know what is going on here? It shows an empty file as the original but I see there is a non-empty file here.

Comment (hppritcha) [reply ompi/mpi/c/profile/Makefile.am]: removed.  a mistake in a rebase.

**Review by hjelmn — APPROVED** [review]

**Comment by hjelmn** [comment]
Given the size of this let's get it in and worry about any issues in follow up commits.

**Comment by gpaulsen** [comment]
Remember we set an arbitrary date of Feb 1st for reviews of this... So interested parties, please take alook.

**Comment by hjelmn** [comment]
I am waiting on this for some coll/sm improvements. While taking a look I remembered i did some work here that will help with that change. Feb 1 sounds good to me.

File: ompi/communicator/communicator.h
Comment (bwbarrett) [inline ompi/communicator/communicator.h:524]: please update docs on what "local" means in this context

File: ompi/datatype/ompi_datatype_module.c
Comment (bwbarrett) [inline ompi/datatype/ompi_datatype_module.c:690]: Is this right?  I can't figure out why you're getting 2 references (one at line 682, one here at line 690).  If it is right, can you add a comment about why it's right?

File: ompi/debuggers/ompi_msgq_dll.c
Comment (bwbarrett) [inline ompi/debuggers/ompi_msgq_dll.c:653]: This needs to be fixed before committing.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: ompi/datatype/ompi_datatype_module.c
Comment (hppritcha) [reply ompi/datatype/ompi_datatype_module.c:690]: looks like a rebase error.  fixed in #9958

**Pull request merged**
